azure/services/tags/tags.go

```diff
@@ -105,13 +105,14 @@ func (s *Service) Reconcile(ctx context.Context) error {
 					return errors.Wrap(err, "cannot update tags")
 				}
 			}
-
-			// We also need to update the annotation if anything changed.
-			if err := s.Scope.UpdateAnnotationJSON(tagsSpec.Annotation, newAnnotation); err != nil {
-				return err
-			}
 			log.V(2).Info("successfully updated tags")
 		}
+
+		// We also need to update the annotation even if nothing changed to
+		// ensure it's set immediately following resource creation.
+		if err := s.Scope.UpdateAnnotationJSON(tagsSpec.Annotation, newAnnotation); err != nil {
+			return err
+		}
```

Comment:
Will this change lead to a new reconciliation for every tag from the for loop up above (as the resource version gets updated on annotation update) ? 
If yes, could such a change create throttling situation?

Reply:
This `UpdateAnnotationJSON` method doesn't make any API calls, it only sets the annotation on the corresponding CAPZ resource like this: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/7c920383f52a58547bd6ac9ae161b553ca564111/azure/scope/cluster.go#L980-L999

Then at the very end of the AzureCluster-level `Reconcile`, that resource is committed to the API with all of the changes made in that reconciliation, so different iterations of this for loop would be coalesced into a single Kubernetes API update. I think this change in particular will only introduce at most one additional reconciliation during the lifetime of the resource which adds the annotation the first time.

Reply:
Thank you for pinning `cluster-api-provider-azure/azure/scope/cluster.go` lines above, helps to put things in perspective. 
Following the code flow from the perspective of generating additional reconciliations, I do not see any `Patch/Update/return reconcile:true` calls on updating annotations or in function `s.Scope.UpdateAnnotationJSON()`. 

So looks good to me!